castervoice/lib/merge/ccrmerging2/transformers/text_replacer/text_replacer.py

```diff
@@ -0,0 +1,154 @@
+from dragonfly.grammar.elements import Choice
+
+from castervoice.lib import printer
+from castervoice.lib.merge.ccrmerging2.transformers.base_transformer import BaseRuleTransformer
+from castervoice.lib.merge.ccrmerging2.transformers.text_replacer.tr_item import TRItem
+from castervoice.lib.merge.ccrmerging2.transformers.text_replacer.tr_parser import TRParser
+
+
+def _preserve(spec):
+    p = TRItem()
+    p.original = spec
+
+    angles_mode = False
+    preserved_word = ""
+    cleaned_spec = ""
+    for c in spec:
+        if c == '<':
+            angles_mode = True
+            cleaned_spec += "<?"
+            continue
+        elif c == '>':
+            angles_mode = False
+            p.extras.append(preserved_word)
+            preserved_word = ""
+            cleaned_spec += ">"
+            continue
+        elif angles_mode:
+            preserved_word += c
+        else:
+            cleaned_spec += c
+    p.cleaned = cleaned_spec
+    p.altered = cleaned_spec
+
+    return p
+
+
+def _restore(pspec):
+    p = pspec
+
+    q = p.altered.split("?")
+    if len(q) == 1:  # no bound extras
+        return p.altered  # so just return the result
+
+    # intersperse the lists
+    c = [b for a in map(None, q, p.extras) for b in a if b is not None]
```

Comment:
This works with python 2 however in python 3 it would lead to a TypeError 'NoneType' object is not callable. I think the logic follows as "The first parameter you pass to the loop function is None but your function is expecting an callable object, which None object isn't."